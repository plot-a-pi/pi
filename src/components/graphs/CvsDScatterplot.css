.container {
  display: flex;
  flex-direction: column;
  align-items: stretch;
  width: 100%;
  height: auto;
  margin: auto;

  @media screen and (min-aspect-ratio: 3/2) and (max-width: 1199px){
    width: 75%;
  }
}

.svg {
  background:whitesmoke;
  display: block;
  width: 100%;
  color: #212E59;
}

.csv {
  height: 0px;
}

