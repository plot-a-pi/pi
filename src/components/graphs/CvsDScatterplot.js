import React, { useRef, useEffect, useState } from 'react';
import { PropTypes } from 'prop-types';
import styles from './CvsDScatterplot.css';
import { select } from 'd3';
import ResizeObserver from 'resize-observer-polyfill';
import CSVButton from '../common/CSVButton';
import { makeCvsDScatterplot } from '../../d3/helpers';

const CvsDScatterplot = ({ data, stats, title, xLabel, yLabel }) => {
  const svgRef = useRef(null);
  const wrapperRef = useRef(null);
  const dataForCSV = data.map(d => ([d.diameter, d.circumference]));

  const useResizeObserver = ref => {
    const [dimensions, setDimensions] = useState(null);

    useEffect(() => {
      const observeTarget = ref.current;
      const resizeObserver = new ResizeObserver(entries => {
        entries.forEach(entry => setDimensions(entry.contentRect));
      });
      resizeObserver.observe(observeTarget);
      return () => {
        resizeObserver.unobserve(observeTarget);
      };
    }, [ref]);
    return dimensions;
  };

  const dimensions = useResizeObserver(wrapperRef);

  useEffect(() => {
    const svg = select(svgRef.current);
    const { width } = dimensions || wrapperRef.current.getBoundingClientRect();
    if(!dimensions) return;

    const wrapper = select(wrapperRef.current);
    wrapper.style('height', `${width * 0.77}px`);

    makeCvsDScatterplot(svg, data, stats, width);

  }, [dimensions, data]);

  return (
    <div className={styles.container} ref={wrapperRef}>
      <svg className={styles.svg} ref={svgRef}>
        <text className={'title'} fill='#212E59'>{title}</text>
        <g className={'x-axis'}></g>
        <text className={'x-label'} fill='#212E59'>{xLabel}</text>
        <g className={'y-axis'}></g>
        <text className={'y-label'} fill='#212E59'>{yLabel}</text>
        <line className={'line'}></line>
      </svg>
      <CSVButton header1='Diameter (in)' header2='Circumference (in)' data={dataForCSV} />  
    </div>
  );
};

CvsDScatterplot.propTypes = {
  data: PropTypes.array.isRequired,
  stats: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  xLabel: PropTypes.string.isRequired,
  yLabel: PropTypes.string.isRequired
};

export default CvsDScatterplot;
