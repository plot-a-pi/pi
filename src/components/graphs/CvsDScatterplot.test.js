import React from 'react';
import { shallow } from 'enzyme';
import CvsDScatterplot from './CvsDScatterplot';

jest.mock('react-socket-io-hooks', () => ({
  useEmitEvent: () => {}, 
  useSocketState: () => ({
    recentUserPoint: {}
  }), 
  useSocket: () => {}
}));

describe('CvsDScatterplot component', () => {
  it('renders CvsDScatterplot', () => {
    const wrapper = shallow(<CvsDScatterplot data={[[1, 2]]} title={'test title'} xLabel={'test x-label'} yLabel={'test y-label'} />);
    expect(wrapper).toMatchSnapshot();
  });
});
