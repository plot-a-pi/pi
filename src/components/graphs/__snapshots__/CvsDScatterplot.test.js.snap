// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvsDScatterplot component renders CvsDScatterplot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CvsDScatterplot
    data={
      Array [
        Object {
          "circumferene": 3.14,
          "diameter": 1,
        },
      ]
    }
    line={true}
    stats={Object {}}
    title="test title"
    xLabel="test x-label"
    yLabel="test y-label"
  />
</ContextProvider>
`;
